src/solver/impls/snes/snes.cxx

```diff
@@ -86,33 +127,457 @@ int SNESSolver::init(int nout, BoutReal tstep) {
   // Set the callback function
   SNESSetFunction(snes, snes_f, FormFunction, this);
 
-  std::string snes_type = (*options)["snes_type"].withDefault("anderson");
+  std::string snes_type = (*options)["snes_type"].withDefault("newtonls");
   SNESSetType(snes, snes_type.c_str());
 
   // Set up the Jacobian
-  MatCreateAIJ(BoutComm::get(), nlocal, nlocal,  // Local sizes
-               PETSC_DETERMINE, PETSC_DETERMINE, // Global sizes
-               3, // Number of nonzero entries in diagonal portion of local submatrix
-               PETSC_NULL,
-               0, // Number of nonzeros per row in off-diagonal portion of local submatrix
-               PETSC_NULL, &Jmf);
+  bool matrix_free =
+      (*options)["matrix_free"].doc("Use matrix free Jacobian?").withDefault<bool>(false);
+  if (matrix_free) {
+    /*
+      PETSc SNES matrix free Jacobian, using a different
+      operator for differencing.
+
+      See PETSc examples
+      http://www.mcs.anl.gov/petsc/petsc-current/src/snes/examples/tests/ex7.c.html
+      and this thread:
+      http://lists.mcs.anl.gov/pipermail/petsc-users/2014-January/020075.html
+
+     */
+    MatCreateSNESMF(snes, &Jmf);
+
+    // Set a function to be called for differencing
+    // This can be a linearised form of the SNES function
+    MatMFFDSetFunction(Jmf, FormFunctionForDifferencing, this);
+
+    // Calculate Jacobian matrix free using FormFunctionForDifferencing
+    SNESSetJacobian(snes, Jmf, Jmf, MatMFFDComputeJacobian, this);
+
+  } else {
+    // Calculate the Jacobian using finite differences
+
+    bool use_coloring = (*options)["use_coloring"]
+                            .doc("Use matrix coloring to calculate Jacobian?")
+                            .withDefault<bool>(true);
+
+    if (use_coloring) {
+      // Use matrix coloring
+      // This greatly reduces the number of times the rhs() function needs
+      // to be evaluated when calculating the Jacobian.
+
+      // Use global mesh for now
+      Mesh* mesh = bout::globals::mesh;
+
+      //////////////////////////////////////////////////
+      // Get the local indices by starting at 0
+      Field3D index = globalIndex(0);
+
+      //////////////////////////////////////////////////
+      // Pre-allocate PETSc storage
+
+      int localN = getLocalN(); // Number of rows on this processor
+      int n2d = f2d.size();
+      int n3d = f3d.size();
+
+      // Set size of Matrix on each processor to localN x localN
+      MatCreate(BoutComm::get(), &Jmf);
+      MatSetSizes(Jmf, localN, localN, PETSC_DETERMINE, PETSC_DETERMINE);
+      MatSetFromOptions(Jmf);
+
+      PetscInt *d_nnz, *o_nnz;
+      PetscMalloc((localN) * sizeof(PetscInt), &d_nnz);
```

Comment:
The [docs suggest using](https://petsc.org/release/docs/manualpages/Sys/PetscMalloc.html) [`PetscNew`](https://petsc.org/release/docs/manualpages/Sys/PetscNew.html#PetscNew) or [`PetscMalloc1`](https://petsc.org/release/docs/manualpages/Sys/PetscMalloc1.html#PetscMalloc1) instead.

Could we even use `std::vector<PetscInt> d_nnz(localN)` here instead? I can't quite get my head around all the alignment business and if it's required, but I think this should be fine? This would eliminate the calls to `PetscFree`, and should hopefully avoid all the warnings about pointer arithmetic